File: Src/API/Information_Technology/Web/Students/student_dashboard.py
from API import settings as st
from flask import Blueprint, request, jsonify
import mysql.connector

student_dashboard_bp = Blueprint('student_dashboard', __name__)


def get_connection():
    return mysql.connector.connect(**st.DB_CONFIG)


# جلب إحصائيات الطالب للبطاقات
@student_dashboard_bp.route('/api/student_dashboard_info', methods=['POST'])
def student_dashboard_info():
    data = request.get_json() or {}
    student_id = data.get('student_id')

    if not student_id:
        return jsonify({"success": False, "message": "الرقم الجامعي مطلوب"}), 400

    try:
        with get_connection() as conn:
            with conn.cursor(dictionary=True) as cursor:
                
                # جلب بيانات الطالب الأساسية
                cursor.execute("""
                    SELECT gpa, completed_hours, academic_year 
                    FROM students 
                    WHERE student_id = %s
                """, (student_id,))
                student = cursor.fetchone()

                if not student:
                    return jsonify({"success": False, "message": "الطالب غير موجود"}), 404

                # جلب جميع العدادات باستعلام واحد
                cursor.execute("""
                    SELECT 
                        (SELECT COUNT(*) FROM registration_requests WHERE student_id = %s AND status = 'معلق') AS pending_count,
                        (SELECT COUNT(*) FROM enrollments WHERE student_id = %s) AS enrolled_count,
                        (SELECT COALESCE(SUM(c.credit_hours), 0) FROM enrollments e 
                         JOIN sections s ON e.section_id = s.section_id 
                         JOIN courses c ON s.course_id = c.course_id 
                         WHERE e.student_id = %s) AS enrolled_hours,
                        (SELECT COUNT(*) FROM student_records WHERE student_id = %s AND final_grade >= 50) AS passed_courses
                """, (student_id, student_id, student_id, student_id))
                
                stats = cursor.fetchone()

        return jsonify({
            "success": True,
            "data": {
                "gpa":             float(student['gpa']),
                "completed_hours": int(student['completed_hours']),
                "academic_year":   int(student['academic_year']),
                "pending_count":   int(stats['pending_count']),
                "enrolled_count":  int(stats['enrolled_count']),
                "enrolled_hours":  int(stats['enrolled_hours']),
                "passed_courses":  int(stats['passed_courses']),
            }
        })

    except mysql.connector.Error as err:
        return jsonify({"success": False, "message": f"خطأ في قاعدة البيانات: {str(err)}"}), 500


# جلب المواد المسجلة للفصل الحالي
@student_dashboard_bp.route('/api/student_dashboard_courses', methods=['POST'])
def student_dashboard_courses():
    data = request.get_json() or {}
    student_id = data.get('student_id')

    if not student_id:
        return jsonify({"success": False, "message": "الرقم الجامعي مطلوب"}), 400

    try:
        with get_connection() as conn:
            with conn.cursor(dictionary=True) as cursor:
                # استعلام جلب تفاصيل المواد والمدرسين
                cursor.execute("""
                    SELECT
                        c.course_id,
                        c.course_name,
                        c.credit_hours,
                        i.full_name  AS instructor_name,
                        s.days,
                        s.start_time,
                        s.end_time,
                        s.room
                    FROM enrollments e
                    JOIN sections    s ON e.section_id    = s.section_id
                    JOIN courses     c ON s.course_id     = c.course_id
                    JOIN instructors i ON s.instructor_id = i.instructors_id
                    WHERE e.student_id = %s
                """, (student_id,))
                rows = cursor.fetchall()

        courses = []
        for row in rows:
            # تنسيق وقت البداية والنهاية
            start = str(row['start_time'])[:5] if row['start_time'] else '—'
            end   = str(row['end_time'])[:5]   if row['end_time']   else '—'
            
            courses.append({
                "course_id":        row['course_id'],
                "course_name":      row['course_name'],
                "credit_hours":     int(row['credit_hours']),
                "instructor_name":  row['instructor_name'],
                "days":             row['days'],
                "time":             f"{start} - {end}",
                "room":             row['room'],
            })

        return jsonify({"success": True, "courses": courses})

    except mysql.connector.Error as err:
        return jsonify({"success": False, "message": f"خطأ في قاعدة البيانات: {str(err)}"}), 500


# جلب الإشعارات الخاصة بالطالب
@student_dashboard_bp.route('/api/student_dashboard_notifications', methods=['POST'])
def student_dashboard_notifications():
    data = request.get_json() or {}
    student_id = data.get('student_id')

    if not student_id:
        return jsonify({"success": False, "message": "الرقم الجامعي مطلوب"}), 400

    try:
        with get_connection() as conn:
            with conn.cursor(dictionary=True) as cursor:
                cursor.execute("""
                    SELECT
                        pn.notification_id,
                        pn.change_type,
                        pn.old_value,
                        pn.new_value,
                        pn.message,
                        pn.sent_at,
                        c.course_name
                    FROM private_notification pn
                    JOIN sections s ON pn.section_id = s.section_id
                    JOIN courses  c ON s.course_id   = c.course_id
                    WHERE pn.student_id = %s
                    ORDER BY pn.sent_at DESC
                    LIMIT 20
                """, (student_id,))
                rows = cursor.fetchall()

        for row in rows:
            if row['sent_at']:
                row['sent_at'] = row['sent_at'].strftime('%Y-%m-%d %H:%M')

        return jsonify({
            "success": True,
            "private": rows,
            "public":  []
        })

    except mysql.connector.Error as err:
        return jsonify({"success": False, "message": f"خطأ في قاعدة البيانات: {str(err)}"}), 500

File: Src/Web/Information_Technology/JS/Students/student_dashboard.js
// student_dashboard.js — منطق لوحة تحكم الطالب وربط البيانات الحية
document.addEventListener("DOMContentLoaded", async () => {

    const currentStudentId   = localStorage.getItem("userId");
    const currentStudentName = localStorage.getItem("userName");

    // 1. التحقق من أمان الجلسة
    if (!currentStudentId) {
        alert("جلسة غير صالحة، يرجى تسجيل الدخول أولاً");
        window.location.href = "/";
        return;
    }

    const inject = (container, html) => {
        container.innerHTML = html;
        container.querySelectorAll('script').forEach(old => {
            const s = document.createElement('script');
            [...old.attributes].forEach(a => s.setAttribute(a.name, a.value));
            s.textContent = old.textContent;
            old.replaceWith(s);
        });
    };

    // 2. جلب ملفات الواجهة المشتركة (التوب بار والسايدبار) وحقنها برمجياً
    try {
        const topbarHeader = document.querySelector(".topbar");
        const sidebarAside = document.querySelector(".sidebar");

        const [topbarHtml, sidebarHtml] = await Promise.all([
            fetch('/Information_Technology/HTML/Students/student_topbar.html').then(r => r.text()),
            fetch('/Information_Technology/HTML/Students/student_sidebar.html').then(r => r.text())
        ]);

        if (topbarHeader) inject(topbarHeader, topbarHtml);
        if (sidebarAside) inject(sidebarAside, sidebarHtml);

    } catch (error) {
        console.error("خطأ أثناء جلب وتحديث المكونات المشتركة:", error);
    }

    // 3. تعبئة اسم الطالب والتاريخ في شريط الميتا بالواجهة الرئيسية
    const metaNameElem = document.getElementById("metaName");
    if (metaNameElem && currentStudentName) {
        metaNameElem.textContent = currentStudentName;
    }

    const metaDateElem = document.getElementById("metaDate");
    if (metaDateElem) {
        metaDateElem.textContent = new Date().toLocaleDateString("ar-JO", {
            year: "numeric", month: "long", day: "numeric"
        });
    }

    // 4. دالة لاقتصاص الأسماء المركبة للترحيب بالطالب
    const getSmartFirstName = (fullName) => {
        if (!fullName) return "الطالب";
        const nameParts = fullName.trim().split(/\s+/);
        const compoundPrefixes = ["عبد", "أبو", "أم", "صلاح", "نور", "سيف", "بهاء", "ذو", "علاء", "شمس", "جمال", "كمال", "محمد"];
        if (compoundPrefixes.includes(nameParts[0]) && nameParts.length > 1) {
            return `${nameParts[0]} ${nameParts[1]}`;
        }
        return nameParts[0];
    };

    const welcomeElem = document.getElementById("welcomeStudentName");
    if (welcomeElem && currentStudentName) {
        welcomeElem.textContent = getSmartFirstName(currentStudentName);
    }

    // 5. جلب إحصائيات البطاقات الحية عبر مسار نسبي مباشر للسيرفر
    fetch('/api/student_dashboard_info', {
        method:  'POST',
        headers: { 'Content-Type': 'application/json' },
        body:    JSON.stringify({ student_id: currentStudentId })
    })
    .then(res => {
        if (!res.ok) throw new Error("فشل في جلب بيانات ملخص الحساب");
        return res.json();
    })
    .then(result => {
        if (!result.success) return;
        const info = result.data;

        const set = (id, val) => {
            const el = document.getElementById(id);
            if (el) el.textContent = val;
        };

        // توزيع البيانات الحية القادمة من قاعدة البيانات على الكروت
        set("studentgpa",      info.gpa);
        set("completedhours",  info.completed_hours);
        set("academicyear",    info.academic_year);
        set("pendingrequests", info.pending_count);
        set("approvedrequests",info.enrolled_count);
        set("enrolledhours",   info.enrolled_hours ?? "—");
        set("passedcourses",   info.passed_courses ?? "—");
        set("remaininghours",  Math.max(0, 132 - info.completed_hours));
    })
    .catch(err => console.error("خطأ أثناء تحديث ملخص الحساب:", err));

    // 6. جلب جدول المواد المسجلة للفصل الحالي عبر مسار نسبي مباشر
    fetch('/api/student_dashboard_courses', {
        method:  'POST',
        headers: { 'Content-Type': 'application/json' },
        body:    JSON.stringify({ student_id: currentStudentId })
    })
    .then(res => {
        if (!res.ok) throw new Error("فشل في جلب المواد المسجلة");
        return res.json();
    })
    .then(result => {
        const tbody = document.getElementById("studentCoursesTable");
        const badge = document.getElementById("coursesCount");
        if (!result.success || !tbody) return;

        const courses = result.courses;

        if (badge) badge.textContent = `${courses.length} مواد مسجلة`;

        if (courses.length === 0) {
            tbody.innerHTML = `<tr><td colspan="6" style="text-align:center;">لا يوجد مواد مسجلة للفصل الحالي</td></tr>`;
            return;
        }

        // بناء صفوف الجدول ديناميكياً بناءً على بيانات الـ MySQL
        tbody.innerHTML = courses.map(c => `
            <tr>
                <td>${c.course_id}</td>
                <td>${c.course_name}</td>
                <td>${c.credit_hours}</td>
                <td>${c.instructor_name}</td>
                <td>${c.days} — ${c.time}</td>
                <td>${c.room}</td>
            </tr>
        `).join("");
    })
    .catch(err => console.error("خطأ أثناء جلب المواد المسجلة:", err));

    // 7. جلب الإشعارات الخاصة بالطالب
    fetch('/api/student_dashboard_notifications', {
        method:  'POST',
        headers: { 'Content-Type': 'application/json' },
        body:    JSON.stringify({ student_id: currentStudentId })
    })
    .then(res => res.json())
    .then(result => {
        if (!result.success) return;
        renderPrivateNotifications(result.private);
    })
    .catch(err => console.error("خطأ أثناء جلب الإشعارات:", err));
});

function renderPrivateNotifications(notifications) {
    const badge = document.getElementById("privateBadge");
    const body  = document.getElementById("privateBody");
    if (!body) return;

    const count = notifications.length;
    if (badge) badge.textContent = `${count} إشعار`;

    if (count === 0) {
        body.innerHTML = `
            <div class="noti-empty-state">
                <div class="empty-check-icon">✓</div>
                <p class="empty-text">لا توجد إشعارات خاصة بالطالب</p>
            </div>`;
        return;
    }

    const icons = { 'وقت': '🕐', 'قاعة': '🏫', 'مدرس': '👨‍🏫', 'يوم': '📅' };

    body.classList.add("has-items");
    body.innerHTML = notifications.map(n => `
        <div class="noti-item">
            <div class="noti-item-header">
                <span class="noti-item-type">
                    ${icons[n.change_type] || '📌'} تغيير ${n.change_type} — ${n.course_name}
                </span>
                <span class="noti-item-date">${n.sent_at}</span>
            </div>
            <p class="noti-item-text">${n.message}</p>
        </div>
    `).join('');
}
